pallets/pools-registry/src/lib.rs

```diff
@@ -0,0 +1,237 @@
+{
+	metadata: BoundedVec<u8, MetaSize>,
+}
+
+type PoolMetadataOf<T> = PoolMetadata<<T as Config>::MaxSizeMetadata>;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		type Balance: Member
+			+ Parameter
+			+ AtLeast32BitUnsigned
+			+ Default
+			+ Copy
+			+ MaxEncodedLen
+			+ FixedPointOperand
+			+ From<u64>
+			+ From<u128>
+			+ TypeInfo
+			+ TryInto<u64>;
+
+		type PoolId: Member
+			+ Parameter
+			+ Default
+			+ Copy
+			+ HasCompact
+			+ MaxEncodedLen
+			+ core::fmt::Debug;
+
+		type CurrencyId: Parameter + Copy;
+
+		type Metadata: Eq
+			+ PartialEq
+			+ Member
+			+ Parameter
+			+ Default
+			+ Copy
+			+ HasCompact
+			+ MaxEncodedLen
+			+ core::fmt::Debug;
+
+		type TrancheId: Member
+			+ Parameter
+			+ Default
+			+ Copy
+			+ MaxEncodedLen
+			+ TypeInfo
+			+ From<[u8; 16]>;
+
+		/// Max size of Metadata
+		#[pallet::constant]
+		type MaxSizeMetadata: Get<u32> + Copy + Member + scale_info::TypeInfo;
+
+		type Permission: Permissions<
+			Self::AccountId,
+			Scope = PermissionScope<Self::PoolId, Self::CurrencyId>,
+			Role = Role<Self::TrancheId, Moment>,
+			Error = DispatchError,
+		>;
+
+		/// Weight Information
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::storage]
+	#[pallet::getter(fn get_pool_metadata)]
+	pub(super) type PoolMetadata<T: Config> =
+		StorageMap<_, Blake2_256, T::PoolId, PoolMetadataOf<T>>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Pool metadata was set.
+		MetadataSet {
+			pool_id: T::PoolId,
+			metadata: BoundedVec<u8, T::MaxSizeMetadata>,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Invalid metadata passed
+		BadMetadata,
+		/// Pre-requirements for a TrancheUpdate are not met
+		/// for example: Tranche changed but not its metadata or vice versa
+		InvalidTrancheUpdate,
+		/// No metada for the given currency found
+		MetadataForCurrencyNotFound,
+		/// No Metadata found for the given PoolId
+		NoSuchPoolMetadata,
+		/// The given tranche token name exceeds the length limit
+		TrancheTokenNameTooLong,
+		/// The given tranche symbol name exceeds the length limit
+		TrancheSymbolNameTooLong,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Sets the IPFS hash for the pool metadata information.
+		///
+		/// The caller must have the `PoolAdmin` role in order to
+		/// invoke this extrinsic.
+		#[pallet::weight(T::WeightInfo::set_metadata(metadata.len().try_into().unwrap_or(u32::MAX)))]
+		pub fn set_metadata(
+			origin: OriginFor<T>,
+			pool_id: T::PoolId,
+			metadata: Vec<u8>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			ensure!(
+				T::Permission::has(
```

Comment:
What should the permissions be here? Is it still tightly coupled with `PoolAdmin`, should we introduce new permissions? @offerijns @mustermeiszer @branan

Reply:
I went for something like

```rust
ensure!(
	T::PreConditions::check(OrderType::Investment {
		who: who.clone(),
		investment_id,
		amount
	}),
	BadOrigin
);
```

whenever I did not want to commit to anything specific in the pallet. Then the runtimes themselves can define what they want to do. I am not sure, if this is the best idea, but its a clean cut, as the registry must not now about roles.